internal/scheduler/legacyscheduler.go

```diff
@@ -95,595 +97,579 @@ func SchedulingConstraintsFromSchedulingConfig(
-	for gang, err := it.queuedGangIterator.Peek(); gang != nil; gang, err = it.queuedGangIterator.Peek() {
+func (sch *Rescheduler) schedule(ctx context.Context, inMemoryJobRepo *InMemoryJobRepository, jobRepo JobRepository) ([]LegacySchedulerJob, error) {
+	queues := make([]*Queue, 0, len(sch.priorityFactorByQueue))
+	for queue, priorityFactor := range sch.priorityFactorByQueue {
+		evictedIt, err := inMemoryJobRepo.GetJobIterator(ctx, queue)
 		if err != nil {
 			return nil, err
 		}
-		if v, ok, err := it.f(gang); err != nil {
+		queueIt, err := NewQueuedJobsIterator(ctx, queue, jobRepo)
+		if err != nil {
 			return nil, err
-		} else if ok {
-			return v, nil
 		}
-		if err := it.queuedGangIterator.Clear(); err != nil {
+		queue, err := NewQueue(
+			queue,
+			priorityFactor,
+			NewMultiJobsIterator(evictedIt, queueIt),
+		)
+		if err != nil {
 			return nil, err
 		}
+		queues = append(queues, queue)
 	}
-	return nil, nil
+	sched, err := NewLegacyScheduler(
+		ctx,
+		sch.constraints,
+		sch.config,
+		sch.nodeDb,
+		queues,
+		sch.usageByQueueAndPriority,
+	)
+	if err != nil {
+		return nil, err
+	}
+	rescheduledJobs, err := sched.Schedule()
+	if err != nil {
+		return nil, err
+	}
+	sched.SchedulingRoundReport.ClearJobSpecs()
+	if sch.schedulingReportsRepository != nil {
+		sch.schedulingReportsRepository.AddSchedulingRoundReport(sched.SchedulingRoundReport)
+	}
+	sch.usageByQueueAndPriority = UpdateUsage(
+		sch.usageByQueueAndPriority,
+		rescheduledJobs,
+		sch.config.Preemption.PriorityClasses,
+		Add,
+	)
+	if s := JobsSummary(rescheduledJobs); s != "" {
+		log.Infof("rescheduled %d jobs; %s", len(rescheduledJobs), s)
+	}
+	return rescheduledJobs, nil
 }
 
-func (it *QueueCandidateGangIterator) f(gang []LegacySchedulerJob) ([]*JobSchedulingReport, bool, error) {
-	if gang == nil {
-		return nil, false, nil
-	}
-	reports, err := it.schedulingReportsFromJobs(it.ctx, gang)
+// For each node in the NodeDb, compare assigned jobs relative to the initial snapshot.
+// Jobs no longer assigned to a node are preemtped.
+// Jobs assigned to a node that weren't earlier are scheduled.
+//
+// Compare the NodeJobDiff with expected preempted/scheduled jobs to ensure NodeDb is consistent.
+// This is only to validate that nothing unexpected happened during scheduling.
+//
+// TODO: Returns a map from job to node until we track that explicitly in Scheduler.Schedule().
+func (sch *Rescheduler) validateSchedulingConsistency(
+	ctx context.Context,
+	snapshot *memdb.Txn,
+	preemptedJobsById,
+	scheduledJobsById map[string]LegacySchedulerJob,
+) (map[string]*schedulerobjects.Node, error) {
+	preempted, scheduled, err := NodeJobDiff(snapshot, sch.nodeDb.Txn(false))
 	if err != nil {
-		return nil, false, err
+		return nil, err
 	}
-	unschedulableReason := ""
-	for _, report := range reports {
-		if report.UnschedulableReason != "" {
-			unschedulableReason = report.UnschedulableReason
-			break
+	for jobId := range preemptedJobsById {
```

Comment:
This check (and  the one below) are a bit weird?  Are we not confident we're scheduling consistently here?

I guess to expand, my rule would be that you only run checks when you can't be sure of something(e.g.input has come from by a user  or you're sending data over a flaky connection). In this case I think all scheduling being done here is (or at least should) be done using pre-validated input using a  deterministic algo we control.

Reply:
Just noticed that this entire function is just runtime checks.  I'd remove the whole thing.

Reply:
Yes, we currently only need it to figure out where jobs ended up (because `Scheduler` doesn't track nodes). I'm changing `Scheduler` to track that explicitly, after which we won't need to call this function. It has been useful to find bugs in staging. But maybe the right trade-off is to put it behind a flag that we leave disabled in prod?

Reply:
I think for now- put it behind a flag and we can turn it on in  staging.  I think we should aim to get rid of it sooner rather than later though because it's basically a band aid.